### PR TITLE
[Train] Add example of pre-training Llama model on Intel Gaudi

### PR DESCRIPTION
**Pull request by harborn**
## Why are these changes needed?

To leverage the potential of Intel Gaudi accelerator, we extend Ray Train's capabilities by adding support for Intel Gaudi (HPU) hardware. This PR include an example for pre-training Llama-7b on multi HPUs.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
(for train team folks to review)

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 1]
```diff
@@ -0,0 +1,568 @@
+{
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:1]: High-level question: seems that it's using deepspeed zero-3 for pre-training. why we also include megatron here?

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+    "        if iteration:\n",
+    "            # passed value is starting step\n",
+    "            iteration -= 1\n",
+    "            args.consumed_train_samples = iteration * args.global_batch_size\n",
+    "            args.consumed_valid_samples = (\n",
+    "                (args.iteration // args.eval_interval) * args.eval_iters * args.global_batch_size\n",
+    "            )\n",
+    "\n",
+    "        # Data loader only on rank 0 of each model parallel group.\n",
+    "        if args.use_dataset_only or mpu.get_tensor_model_parallel_rank() == 0:\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:189]: Where did we configure the tensor and pipeline parallel group size?

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+    "(RayTrainWorker pid=339380) {'loss': nan, 'grad_norm': nan, 'learning_rate': 4.9250000000000004e-05, 'epoch': 0.0, 'memory_allocated (GB)': 40.42, 'max_memory_allocated (GB)': 93.68, 'total_memory_available (GB)': 94.62}\n",
+    "(RayTrainWorker pid=339380) {'loss': nan, 'grad_norm': nan, 'learning_rate': 4.9e-05, 'epoch': 0.0, 'memory_allocated (GB)': 40.42, 'max_memory_allocated (GB)': 93.68, 'total_memory_available (GB)': 94.62}\n",
+    "(RayTrainWorker pid=339380) {'loss': nan, 'grad_norm': nan, 'learning_rate': 4.875e-05, 'epoch': 0.0, 'memory_allocated (GB)': 40.4, 'max_memory_allocated (GB)': 93.68, 'total_memory_available (GB)': 94.62}\n",
+    "(RayTrainWorker pid=339380) {'loss': nan, 'grad_norm': nan, 'learning_rate': 4.85e-05, 'epoch': 0.0, 'memory_allocated (GB)': 40.4, 'max_memory_allocated (GB)': 93.68, 'total_memory_available (GB)': 94.62}\n",
+    "(RayTrainWorker pid=339380) {'loss': nan, 'grad_norm': nan, 'learning_rate': 4.825e-05, 'epoch': 0.0, 'memory_allocated (GB)': 40.45, 'max_memory_allocated (GB)': 93.68, 'total_memory_available (GB)': 94.62}\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:549]: seems that the loss and grad norm are `nan`, can you try to fix this?

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+    "    if execution_mode == \"eager.compile\":\n",
+    "        pretrain_config[\"training_args\"].update({\"torch_compile_backend\": \"hpu_backend\"})\n",
+    "\n",
+    "    scaling_config = ScalingConfig(num_workers=num_workers,\n",
+    "                                   use_gpu=False,\n",
+    "                                   resources_per_worker={\"CPU\": 1, \"HPU\": 1})\n",
+    "\n",
+    "    # Set backend to hccl in TorchConfig\n",
+    "    torch_config = TorchConfig(backend=\"hccl\")\n",
+    "    runtime_env = {\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:466]: Do we need this if it's empty

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    "def load_datasets(config):\n",
+    "    dataset = MegatronDataset()\n",
+    "    return dataset(config)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Process dataset to dataloader"
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:163]: In general, can we introduce the model sharding layouts, so that the users can better understand why the dataloader is defined this?

Including pp/tp/dp group size, global batch size, per device batch size.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 1 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:1]: this example import megatron just for data processing

Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:1]: Hi, I removed the megatron dependency here, just using deepspeed to pre-train model.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 5 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb]: As I removed megatron dependency, so no PP/TP/DP supported

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 3 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb]: OK, fixed this bug, due to the dataset not processed correctly.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 2 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb]: tensor and pipeline parallel not supported yet, for this example no longer relies on megatron.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+    "(RayTrainWorker pid=289322) {'loss': 6.7815, 'grad_norm': 3.993046522140503, 'learning_rate': 4.633911368015414e-05, 'epoch': 0.22, 'memory_allocated (GB)': 27.32, 'max_memory_allocated (GB)': 93.29, 'total_memory_available (GB)': 94.62}\n",
+    "(RayTrainWorker pid=289322) {'loss': 6.8765, 'grad_norm': 2.5143563747406006, 'learning_rate': 4.624277456647399e-05, 'epoch': 0.23, 'memory_allocated (GB)': 27.34, 'max_memory_allocated (GB)': 93.29, 'total_memory_available (GB)': 94.62}\n",
+    "\n",
+    "...\n",
+    "```"
+   ]
+  }
+ ],
+ "metadata": {
+  "language_info": {
```
Comment (woshiyyya) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:594]:
```suggestion
  "orphan": true,
  "language_info": {
```

**Review by woshiyyya — APPROVED** [review]
Thanks for the update, pretty clean code snippet now!

@justinvyu can you take a look and merge it after the CI passed? Thank you!

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 7]
```diff
@@ -0,0 +1,601 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Llama model pre-training on HPU\n",
```
Comment (carsonwang) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:7]:
```suggestion
    "# Llama model pre-training on Intel Gaudi\n",
```

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+    "\n",
+    "In this Jupyter notebook, we will pre-train a [huggyllama/llama-7b](https://huggingface.co/huggyllama/llama-7b) model by using HPU.\n",
```
Comment (carsonwang) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:9]:
```suggestion
    "In this Jupyter notebook, we will pre-train a [huggyllama/llama-7b](https://huggingface.co/huggyllama/llama-7b) model by using Intel Gaudi accelerators.\n",
```

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 7 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb]: updated.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 8 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb]: updated.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+    "\n",
+    "Next, follow [Run Using Containers](https://docs.habana.ai/en/latest/Installation_Guide/Bare_Metal_Fresh_OS.html?highlight=installer#run-using-containers) to install the Habana drivers and container runtime.\n",
+    "\n",
+    "### Get docker image\n",
+    "``` bash\n",
+    "docker pull vault.habana.ai/gaudi-docker/1.15.1/ubuntu22.04/habanalabs/pytorch-installer-2.2.0:latest\n",
+    "```\n",
+    "### Run docker image\n",
+    "``` bash\n",
+    "docker run -it --runtime=habana -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --net=host --ipc=host vault.habana.ai/gaudi-docker/1.15.1/ubuntu22.04/habanalabs/pytorch-installer-2.2.0:latest\n",
```
Comment (jerome-habana) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:40]: nit - might be work to have the image as part of macro and update as required across releases

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+    "# maybe should mapping your workspace volumns\n",
+    "```\n",
+    "### Install dependency\n",
+    "``` bash\n",
+    "# \"optimum-habana>1.11.1\" if exection mode \"eager\" or \"eager.compile\" \n",
+    "# \"ray>=2.20.0\"\n",
+    "pip install ray[train] notebook transformers datasets evaluate peft accelerate scikit-learn optimum-habana\n",
+    "\n",
+    "# install deepspeed\n",
+    "pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.15.0\n",
```
Comment (jerome-habana) [inline doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:50]: ditto

**Comment by justinvyu** [comment]
@harborn Can you add the orphan tag like we did in the previous PR to pass the CI?

https://github.com/ray-project/ray/pull/44667/files#diff-21132e4fa5d8a49af65d457534637f53c79c76dbd91b945b670e99d3163d9ea4R576

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 10 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:51]: Updated a comment in the notebook.

File: doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb (hunk 9 above)
Comment (harborn) [reply doc/source/train/examples/intel_gaudi/llama_pretrain.ipynb:41]: Updated a comment in the notebook for more information about docker image version

**Comment by harborn** [comment]
> @harborn Can you add the orphan tag like we did in the previous PR to pass the CI?
> 
> https://github.com/ray-project/ray/pull/44667/files#diff-21132e4fa5d8a49af65d457534637f53c79c76dbd91b945b670e99d3163d9ea4R576

done
